Added a custom SASL config option to the Topic Operator #5473 (#9606)
strimzi/strimzi-kafka-operator: documentation/modules/deploying/proc-deploy-topic-operator-standalone.adoc

Reviewer: I think we should add something in the intro to the procedure, with maybe a link too.
Something like: 

````
If you need to configure custom SASL authentication, you can define the necessary authentication properties using the `STRIMZI_SASL_CUSTOM_CONFIG_JSON` environment variable for the standalone operator. 
For example, this configuration may be used for accessing a Kafka cluster in a cloud provider with a custom login module like the link:https://github.com/aws/aws-msk-iam-auth[Amazon MSK Library for AWS Identity and Access Management^] (`aws-msk_iam-auth`).
````

Reviewer: I still think we should add something to the intro like this^

Author: I thought we already put that in with one of the other suggested changes.? I'll get it in.

Author: I updated the docs with that.

<3> The keystore contains the private key for mTLS authentication.
<4> The password for accessing the keystore.

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module like `AWS_MSK_IAM`. The `STRIMZI_SASL_CUSTOM_CONFIG_JSON` is the mechanism to do that.
> The property `STRIMZI_ALTERABLE_TOPIC_CONFIG` defaults to `ALL`. When a subset of properties is alterable, use a comma separated list, as below. In the case that a managed Kafka service doesn't allow any `.spec.config` properties to be set in the `KafkaTopic` a value of `NONE` is equivalent to an empty property.
> Otherwise, a comma separated list of the properties that can be altered is used.

Reviewer 2: Nit: we have this convention of one sentence per line, which helps to make the doc review easier. It will still be rendered as before.

```suggestion
. If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module like `AWS_MSK_IAM`.
The `STRIMZI_SASL_CUSTOM_CONFIG_JSON` is the mechanism to do that.
The property `STRIMZI_ALTERABLE_TOPIC_CONFIG` defaults to `ALL`.
When a subset of properties is alterable, use a comma separated list, as below.
In the case that a managed Kafka service doesn't allow any `.spec.config` properties to be set in the `KafkaTopic` a value of `NONE` is equivalent to an empty property.
Otherwise, a comma separated list of the properties that can be altered is used.
```

<3> The keystore contains the private key for mTLS authentication.
<4> The password for accessing the keystore.

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module like `AWS_MSK_IAM`.
> The `STRIMZI_SASL_CUSTOM_CONFIG_JSON` is the mechanism to do that.

Reviewer: ```suggestion
. If required, configure custom SASL authentication using the STRIMZI_SASL_CUSTOM_CONFIG_JSON environment variable. 
This allows you to specify the necessary SASL properties for authentication.
+
```

Reviewer: also adds a `+` for some space.

> The property `STRIMZI_ALTERABLE_TOPIC_CONFIG` defaults to `ALL`.
> When a subset of properties is alterable, use a comma separated list, as below.
> In the case that a managed Kafka service doesn't allow any `.spec.config` properties to be set in the `KafkaTopic` a value of `NONE` is equivalent to an empty property.
> Otherwise, a comma separated list of the properties that can be altered is used.

Reviewer: ```suggestion
The property `STRIMZI_ALTERABLE_TOPIC_CONFIG` defaults to `ALL`, allowing all `.spec.config` properties to be set in the `KafkaTopic` resource.  
If this setting is not suitable for a managed Kafka service, do as follows:
+
* If only a subset of properties is configurable, list them as comma-separated values.
* If no  properties are to be configured, use `NONE`, which is equivalent to an empty property list.
```

+

> *NOTE:  Only SASL properties can be set with the `STRIMZI_SASL_CUSTOM_CONFIG_JSON` mechanism.*

Reviewer: ```suggestion
NOTE:  Only Kafka configuration properties starting with `sasl.` can be set with the `STRIMZI_SASL_CUSTOM_CONFIG_JSON` environment variable.
```

+
.Example custom SASL configuration
[source,shell,subs=+quotes]
----
# ....
env:
  - name: STRIMZI_SASL_ENABLED
    value: "true"
  - name: STRIMZI_SECURITY_PROTOCOL
    value: SASL_SSL
  - name: STRIMZI_SKIP_CLUSTER_CONFIG_REVIEW # <1>
    value: "true"
  - name: STRIMZI_ALTERABLE_TOPIC_CONFIG # <2>
    value: compression.type, max.message.bytes, message.timestamp.difference.max.ms, message.timestamp.type, retention.bytes, retention.ms
  - name: STRIMZI_SASL_CUSTOM_CONFIG_JSON # <3>
    value: |
      {
        "sasl.mechanism": "AWS_MSK_IAM",
        "sasl.jaas.config": "software.amazon.msk.auth.iam.IAMLoginModule required;",
        "sasl.client.callback.handler.class": "software.amazon.msk.auth.iam.IAMClientCallbackHandler"
      }
  - name: STRIMZI_PUBLIC_CA
    value: "true"
  - name: STRIMZI_TRUSTSTORE_LOCATION
    value: /etc/pki/java/cacerts
  - name: STRIMZI_TRUSTSTORE_PASSWORD
    value: changeit
  - name: STRIMZI_KAFKA_BOOTSTRAP_SERVERS

>     value: your-instance-here.kafka-serverless.us-east-1.amazonaws.com:9098

Reviewer: ```suggestion
    value: my-kafka-cluster-.kafka-serverless.us-east-1.amazonaws.com:9098
```

# ...
----

> <1> Some managed Kafka services don't allow cluster config lookup, so we disable those checks.

Reviewer: ```suggestion
<1> Disables cluster configuration lookup for managed Kafka services that don't allow topic configuration changes.
```

> <2> Some managed Kafka services limit what properties can be updated, so we use this to define what values can be altered.

Reviewer: ```suggestion
<2> Defines the topic configuration properties that can be updated based on the limitations set by managed Kafka services.
```

> <3> JSON with the SASL properties to be set. Only properties beginning with `sasl.` are allowed.

Reviewer: ```suggestion
<3> Specifies the SASL properties to be set in JSON format. Only properties starting with `sasl.` are allowed.
```

+
.Example Dockerfile with external jars
[source,shell,subs=+quotes]
----
FROM strimzi/operator:latest

USER root

RUN mkdir -p ${STRIMZI_HOME}/external-libs
RUN chmod +rx ${STRIMZI_HOME}/external-libs

COPY ./aws-msk-iam-auth-and-dependencies/* ${STRIMZI_HOME}/external-libs/

> ENV  JAVA_CLASSPATH=${STRIMZI_HOME}/external-libs/*

Reviewer 3: ```suggestion
ENV JAVA_CLASSPATH=${STRIMZI_HOME}/external-libs/*
```

<3> The keystore contains the private key for mTLS authentication.
<4> The password for accessing the keystore.

> . If required, configure custom SASL authentication using the STRIMZI_SASL_CUSTOM_CONFIG_JSON environment variable. 

Reviewer: ```suggestion
. If required, configure custom SASL authentication using the `STRIMZI_SASL_CUSTOM_CONFIG_JSON` environment variable. 
```

This allows you to specify the necessary SASL properties for authentication.
+
The property `STRIMZI_ALTERABLE_TOPIC_CONFIG` defaults to `ALL`, allowing all `.spec.config` properties to be set in the `KafkaTopic` resource.  
If this setting is not suitable for a managed Kafka service, do as follows:

> +
> * If only a subset of properties is configurable, list them as comma-separated values.
> * If no  properties are to be configured, use `NONE`, which is equivalent to an empty property list.
> +

Reviewer: ```suggestion
+
--
* If only a subset of properties is configurable, list them as comma-separated values.
* If no  properties are to be configured, use `NONE`, which is equivalent to an empty property list.
--
+
```

Reviewer: This stops the example being pulled inline with the bullets

<3> The keystore contains the private key for mTLS authentication.
<4> The password for accessing the keystore.

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Author: I expected that if you need AWS IAM or AZURE, etc. you would create an image with those so we don't have to include every possible jar. I would like an standard mechanism for external jars other than a new image, but that may be a different conversation.

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Author: Sorry, I don't know what you mean by BTO?

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Reviewer 2: > I would like an standard mechanism for external jars other than a new image, but that may be a different conversation.

As you say, we can discuss that in a separate thread, but for now, we should include a Dockerfile example for this specific use case, so that other users can easily adapt to their needs.

> Sorry, I don't know what you mean by BTO?

Bidirectional Topic Operator, it's the old deprecated implementation. In the latest Strimzi release you have the UTO (Unidirectional Topic Operator) enabled by default. https://strimzi.io/blog/2023/11/02/unidirectional-topic-operator

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Author: Roger that, I get that in Dockerfile example.

Ah, I see, I'll get on that.

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Author: I've run into a problem on MSK Serverless. BatchingTopicController is trying to find out if auto create topics is on. So a pretty reasonable question, but MSK Serverless doesn't expose that info through the usual Kafka mechanism. :/

**Does it even make sense to find pluggable solution for this?**

`Exception in thread "main" java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.InvalidRequestException: Amazon MSK Serverless doesn't support fetching broker or broker-logger configurations.
        at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
        at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:165)
        at io.strimzi.operator.topic.v2.BatchingTopicController.<init>(BatchingTopicController.java:108)
        at io.strimzi.operator.topic.v2.TopicOperatorMain.<init>(TopicOperatorMain.java:70)
        at io.strimzi.operator.topic.v2.TopicOperatorMain.operator(TopicOperatorMain.java:152)
        at io.strimzi.operator.topic.v2.TopicOperatorMain.main(TopicOperatorMain.java:146)
Caused by: org.apache.kafka.common.errors.InvalidRequestException: Amazon MSK Serverless doesn't support fetching broker or broker-logger configurations.`

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Reviewer 4: > I expected that if you need AWS IAM or AZURE, etc. you would create an image with those so we don't have to include every possible jar.

You should explain this in the documentation. Right now someone reading it might not guess all the steps you're expecting them to make.

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Reviewer 4: > `org.apache.kafka.common.errors.InvalidRequestException: Amazon MSK Serverless doesn't support fetching broker or broker-logger configurations`

Perhaps one way to handle this would be to allow the skipping of this check with an env var, e.g. `ASSUME_AUTO_CREATE_TOPICS=FALSE`. If this was not set we'd do the check which we currently do. If it was `=FALSE` then we just trust that the person configuring the operator is correct and skip the check. And in the unlikely case where it's `=TRUE` we fail in the same way as if we'd done the check ourselves.

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Author: Sounds like a plan if you all are good with it. Should I proceed as such?

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Author: Just a note. It works perfectly for normal "Provisioned" MSK clusters using IAM.

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Author: On another note, the details of a working example might be a lot for the docs. What do you all think of a separate doc to explain the AWS example with a link in the Standalone doc?

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Author: This is a repo I did to show the example. [MSK-IAM-Example](https://github.com/john-mcpeek/msk-iam-example/tree/master)

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Reviewer 2: I think a generic Dockerfile example with <placeholders> is enough. As mentioned before, we should also update the BTO documentation. Alternatively, we can only implement and document this feature for the UTO, as the BTO is going away in two releases.

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Reviewer 3: Actually, BTO is scheduled for removal in 0.41. So unless we change it it will be gone in few weeks.

> . If you need a custom `SASL` configuration. For example, to access a cluster in a cloud provider with a custom login module, the `custom` `SASL_MECHANISM` allows for `SASL_CUSTOM_CONFIG`. +

Reviewer 2: BTO is gone in the meantime, so less work to do :)

*NOTE:  Only SASL properties can be set with the `custom` mechanism.*
+
.Example custom SASL configuration
[source,shell,subs=+quotes]
----
# ....

> env:

Author: Roger that.

> env:

Author: I've been busy with other stuff for a while. I'm back on this.

  - name: SASL_MECHANISM # <1>

>     value: custom

Reviewer 4: Thinking about it, I'm a little wary of indicating the custom situation like this. Right now this is very clearly expected to be a recognised SASL mechanism. Technically the names of SASL mechanisms are defined by IANA (https://www.iana.org/assignments/sasl-mechanisms/sasl-mechanisms.xhtml), so we'd be trespassing into their namespace if we allowed `custom`.

But I don't think we need to supply `SASL_MECHANISM` _at all_ in the case you want to support: We can make `SASL_MECHANISM` mutually exclusive with `SASL_CUSTOM_CONFIG_JSON`.

Reviewer 2: I was thinking along the same line.

>   - name: SASL_CUSTOM_CONFIG # <2>

Reviewer 4: 1. I wonder if `SASL_CUSTOM_CONFIG_JSON` would be clearer. 
2. But also, did you consider using `Properties` format rather than JSON? Users would likely be familiar with this because it's what the Kafka Java client's use, so it might avoid them having to convert something they already have as `Properties` format into JSON.

Author: 1. I like adding JSON to the name.
2. JSONs easy to parse and I was worried about how it could be a mess getting people to understand the value as properties, but taking properties shouldn't be a big change if you would prefer that.

Reviewer 4: > I was worried about how it could be a mess getting people to understand the value as properties, but taking properties shouldn't be a big change if you would prefer that.

That's a good point. The case of accidentally using folded style block strings (`>`), like this, bothers me

```yaml
- name: SASL_CUSTOM_CONFIG_PROPERTIES
  value: >
    sasl.mechanism: AWS_MSK_IAM
    sasl.jaas.config: software.amazon.msk.auth.iam.IAMLoginModule required;
    sasl.client.callback.handler.class: software.amazon.msk.auth.iam.IAMClientCallbackHandler
```

That's legal YAML and legal `.properties` format, but it's not what the user intended because it's just the single key `sasl.mechanism` with a malformed value made up of the real value and the remaining keys and values. Kafka itself would catch that when/if there was sufficient validation on the first property defined. I think there's a similar issue with plain flow scalar strings. (I guess this kind of problem is worse than other common YAML problems, like writing `foo: true` [where the value is expected to be a string not a boolean, so needs to be quoted] precisely because in this case the expected and actual types are the same, so it's less likely to be caught.)

So I agree, on balance, that although the JSON is ugly it's less likely to be parsed wrongly. 

Can we though:
* Call it `SASL_CUSTOM_CONFIG_JSON`
* Configure the `ObjectMapper` to support comments. It's non-standard, but I think people might appreciate the possibility to comment on why/how they're setting individual properties.

Author: I'm on it.

Author: This is what I'm working on now. Is `SASL_CUSTOM_CONFIG_JSON` and not having a custom type the solution we are going with?

Reviewer 4: @john-mcpeek yes, I think that's what we landed on, right @fvaleri ?

    value: |
      {
        "sasl.mechanism": "AWS_MSK_IAM",
        "sasl.jaas.config": "software.amazon.msk.auth.iam.IAMLoginModule required;",
        "sasl.client.callback.handler.class": "software.amazon.msk.auth.iam.IAMClientCallbackHandler"
      }
# ...
----
<1> Setting the `custom` SASL mechanism. This enables the custom configuration, so the actual `sasl.mechanism` needs to be included in the `SASL_CUSTOM_CONFIG` value.

> <2> JSON with the SASL properties to be set.

Reviewer 4: ```suggestion
<2> JSON with the SASL properties to be set. Only properties beginning with `sasl.` are allowed.
```
